File: ruby/test_pack.rb
require 'msgpack'

def check(data)
	puts "---"
	pack = data.to_msgpack
	p data
	puts pack.unpack('C*').map{|x|"%02x"%x}.join(' ')
	re = MessagePack::unpack(pack)
	if re != data
		p re
		puts "** TEST FAILED **"
	end
end

check 0
check 1
check 127
check 128
check 255
check 256
check 65535
check 65536
check -1
check -128
check -129
check -32768
check -32769

check 1.0

check ""
check "a"
check "a"*31
check "a"*32

check nil
check true
check false

check []
check [[]]
check [[], nil]

check( {nil=>0} )

check (1<<23)
__END__

ary = []
i = 0
while i < (1<<16)
	ary << i
	i += 1
end
check ary

File: python/test3/test_pack.py
#!/usr/bin/env python
# coding: utf-8

from nose import main
from nose.tools import *

from msgpack import packs, unpacks, Unpacker, Packer

from io import BytesIO

def check(data):
    re = unpacks(packs(data))
    assert_equal(re, data)

def testPack():
    test_data = [
            0, 1, 127, 128, 255, 256, 65535, 65536,
            -1, -32, -33, -128, -129, -32768, -32769,
            1.0,
        b"", b"a", b"a"*31, b"a"*32,
        None, True, False,
        (), ((),), ((), None,),
        {None: 0},
        (1<<23),
        ]
    for td in test_data:
        check(td)

def testPackUnicode():
    test_data = [
        "", "abcd", ("defgh",), "Русский текст",
        ]
    for td in test_data:
        re = unpacks(packs(td, encoding='utf-8'), encoding='utf-8')
        assert_equal(re, td)
        packer = Packer(encoding='utf-8')
        data = packer.pack(td)
        re = Unpacker(BytesIO(data), encoding='utf-8').unpack()
        assert_equal(re, td)

def testPackUTF32():
    test_data = [
        "", "abcd", ("defgh",), "Русский текст",
        ]
    for td in test_data:
        re = unpacks(packs(td, encoding='utf-32'), encoding='utf-32')
        assert_equal(re, td)

def testPackBytes():
    test_data = [
        b"", b"abcd", (b"defgh",),
        ]
    for td in test_data:
        check(td)

def testIgnoreUnicodeErrors():
    re = unpacks(packs(b'abc\xeddef'),
        encoding='utf-8', unicode_errors='ignore')
    assert_equal(re, "abcdef")

@raises(UnicodeDecodeError)
def testStrictUnicodeUnpack():
    unpacks(packs(b'abc\xeddef'), encoding='utf-8')

@raises(UnicodeEncodeError)
def testStrictUnicodePack():
    packs("abc\xeddef", encoding='ascii', unicode_errors='strict')

def testIgnoreErrorsPack():
    re = unpacks(packs("abcФФФdef", encoding='ascii', unicode_errors='ignore'), encoding='utf-8')
    assert_equal(re, "abcdef")

@raises(TypeError)
def testNoEncoding():
    packs("abc", encoding=None)

def testDecodeBinary():
    re = unpacks(packs("abc"), encoding=None)
    assert_equal(re, b"abc")

if __name__ == '__main__':
    main()
